File: impl-csharp/src/Verify.cs
using System.Security.Cryptography;
using System.Text;

namespace NoaReceipt;

public enum VerifyStatus
{
    Valid,
    Unverified,
    Tampered,
    Malformed,
    Untrusted,
}

public readonly struct VerifyResult
{
    public readonly VerifyStatus Status;
    public readonly string Detail;
    public VerifyResult(VerifyStatus status, string detail)
    {
        Status = status;
        Detail = detail;
    }
}

/// <summary>
/// Receipt-chain verification — verdict-equivalent to impl-py/noa_verify.py verify_chain and the
/// TS reference: structural validation, hash-chain integrity, Ed25519 signatures, per-agent key
/// continuity, identity binding (UNTRUSTED), and checkpoint tail-truncation + §5b opener-binding.
/// </summary>
public static class Verifier
{
    private static readonly byte[] ReceiptDomain = Encoding.UTF8.GetBytes("NOA-Receipt-v0.1-sig:");
    private static readonly byte[] CheckpointDomain = Encoding.UTF8.GetBytes("NOA-Checkpoint-v0.1-sig:");
    private const long SafeIntMax = 9007199254740991L;

    private static readonly HashSet<string> CheckpointKeys =
        new(StringComparer.Ordinal) { "spec", "chain", "highestSeq", "headHash", "ts", "sig" };

    // ── hash inputs ──────────────────────────────────────────────────────────
    private static string Sha256Prefixed(string canonical) =>
        "sha256:" + ToHex(SHA256.HashData(Encoding.UTF8.GetBytes(canonical)));

    private static byte[] Sha256Digest(string canonical) =>
        SHA256.HashData(Encoding.UTF8.GetBytes(canonical));

    /// <summary>JCS(receipt WITHOUT chain.hash AND sig.value).</summary>
    private static string ReceiptHashInput(JObj receipt)
    {
        JObj clone = receipt.DeepCloneObj();
        if (clone.Get("chain") is JObj ch) ch.Remove("hash");
        if (clone.Get("sig") is JObj sg) sg.Remove("value");
        return Jcs.Canonicalize(clone);
    }

    /// <summary>JCS(checkpoint WITHOUT sig.value).</summary>
    private static string CheckpointHashInput(JObj cp)
    {
        JObj clone = cp.DeepCloneObj();
        if (clone.Get("sig") is JObj sg) sg.Remove("value");
        return Jcs.Canonicalize(clone);
    }

    // ── checkpoint verification (mirrors _verify_checkpoint) ───────────────────
    private enum CpVerdict { Ok, Unverified, Bad }

    private static CpVerdict VerifyCheckpoint(JVal cpVal, JObj? keyring)
    {
        if (cpVal is not JObj cp) return CpVerdict.Bad;
        foreach (string k in cp.Keys)
            if (!CheckpointKeys.Contains(k)) return CpVerdict.Bad;
        if (StrOf(cp, "spec") != "noa.checkpoint/0.1") return CpVerdict.Bad;
        if (!(cp.Get("chain") is JStr chn && chn.Value.Length > 0)) return CpVerdict.Bad;
        if (cp.Get("highestSeq") is not JInt hs || hs.Value < 0 || hs.Value > SafeIntMax) return CpVerdict.Bad;
        if (!(cp.Get("headHash") is JStr hh && Schema.HashFormat(hh.Value))) return CpVerdict.Bad;
        if (!(cp.Get("ts") is JStr ts && Schema.Rfc3339Instant(ts.Value))) return CpVerdict.Bad;

        if (cp.Get("sig") is not JObj sig) return CpVerdict.Bad;
        foreach (string k in sig.Keys)
            if (k != "alg" && k != "kid" && k != "value") return CpVerdict.Bad;
        if (StrOf(sig, "alg") != "ed25519") return CpVerdict.Bad;
        if (!(sig.Get("kid") is JStr kid && kid.Value.Length > 0)) return CpVerdict.Bad;
        if (!(sig.Get("value") is JStr sval && sval.Value.Length > 0)) return CpVerdict.Bad;

        JVal? pub = keyring?.Get(kid.Value);
        if (pub is not JStr pubStr) return CpVerdict.Unverified;

        try
        {
            byte[] digest = Sha256Digest(CheckpointHashInput(cp));
            byte[] msg = Concat(CheckpointDomain, digest);
            byte[] rawKey = Crypto.SpkiToRaw(pubStr.Value);
            byte[] sigBytes = Crypto.StrictB64Decode(sval.Value);
            return Crypto.Ed25519Verify(rawKey, msg, sigBytes) ? CpVerdict.Ok : CpVerdict.Bad;
        }
        catch (Exception)
        {
            return CpVerdict.Bad;
        }
    }

    private static bool Authorized(JObj manifest, string? agentId, string kid)
    {
        if (agentId is null) return false;
        if (manifest.Get(agentId) is not JArr kids) return false;
        foreach (JVal k in kids.Items)
            if (k is JStr ks && ks.Value == kid) return true;
        return false;
    }

    // ── the chain verifier ─────────────────────────────────────────────────────
    public static VerifyResult VerifyChain(JVal receiptsVal, JVal? keyringVal, JVal? identityVal, JVal? checkpointVal)
    {
        if (receiptsVal is not JArr arr || arr.Items.Count == 0)
            return new VerifyResult(VerifyStatus.Malformed, "input is not a non-empty array");

        if (keyringVal is not null && keyringVal is not JObj)
            return new VerifyResult(VerifyStatus.Malformed, "keyring must be an object (kid -> base64 SPKI)");

        JObj? identity = null;
        if (identityVal is not null)
        {
            if (identityVal is not JObj im)
                return new VerifyResult(VerifyStatus.Malformed, "identityManifest must be an object (agent.id -> kid[])");
            foreach (string aid in im.Keys)
            {
                if (im.Get(aid) is not JArr kids || !kids.Items.TrueForAll(x => x is JStr))
                    return new VerifyResult(VerifyStatus.Malformed, $"identityManifest[\"{aid}\"] must be an array of kid strings");
            }
            identity = im;
        }

        // Step 1: structural validation of every element, BEFORE any hashing.
        for (int idx = 0; idx < arr.Items.Count; idx++)
        {
            var (ok, errs) = Schema.Validate(arr.Items[idx]);
            if (!ok)
                return new VerifyResult(VerifyStatus.Malformed, $"receipt[{idx}]: " + string.Join("; ", errs));
        }

        JObj? keyring = keyringVal as JObj;
        bool haveKeyring = keyringVal is not null;

        var receipts = new List<JObj>();
        foreach (JVal v in arr.Items) receipts.Add((JObj)v);

        string? chainId = ChainOf(receipts[0]);

        // Single chain partition + duplicate-seq detection.
        var bySeq = new Dictionary<long, JObj>();
        foreach (JObj r in receipts)
        {
            if (ChainOf(r) != chainId)
                return new VerifyResult(VerifyStatus.Tampered, "multiple chain partitions");
            long seq = SeqOf(r);
            if (bySeq.ContainsKey(seq))
                return new VerifyResult(VerifyStatus.Tampered, $"duplicate seq {seq}");
            bySeq[seq] = r;
        }

        // Chain-wide scope.tenant consistency (fail-closed; matches the TS reference DEFAULT,
        // impl-py, impl-go and impl-rust). scope.tenant sits beside scope.chain but was enforced
        // nowhere, so a chain mixing tenants verified clean in every implementation. Tenant
        // isolation is a security boundary; the partition split's verdict class is the right one.
        // Walked in SEQ order; a MISSING tenant is distinct from a present one.
        // Only a present->DIFFERENT-present drift is a cross-tenant splice. absent<->present is
        // a producer-version change on an OPTIONAL field, not tampering (see src/verify.ts).
        //
        // The comparison is NOT adjacent: comparing neighbours let an omission RESET the boundary,
        // so `acme -> globex` was TAMPERED while `acme -> absent -> globex` — the same splice with
        // one optional field left out in between — was VALID in all five implementations. The last
        // PRESENT tenant is carried across absences instead. Absence stays non-fatal; it just no
        // longer erases what the chain already committed to.
        string? lastPresentTenant = null;
        for (int s = 0; s < receipts.Count; s++)
        {
            if (!bySeq.TryGetValue(s, out JObj? curT))
                break; // a gap is reported by the seq-walk below; do not pre-empt it
            string? cur = TenantOf(curT);
            if (cur is null)
                continue; // absence never resets the boundary, and is never fatal
            if (lastPresentTenant is not null && !string.Equals(cur, lastPresentTenant, StringComparison.Ordinal))
                return new VerifyResult(VerifyStatus.Tampered, "cross-tenant splice");
            lastPresentTenant = cur;
        }

        var pinned = new Dictionary<string, string>(StringComparer.Ordinal);
        JObj? prev = null;
        for (int s = 0; s < receipts.Count; s++)
        {
            if (!bySeq.TryGetValue(s, out JObj? r))
                return new VerifyResult(VerifyStatus.Tampered, $"seq gap: missing {s}");

            string hi;
            try
            {
                hi = ReceiptHashInput(r);
            }
            catch (Exception e)
            {
                return new VerifyResult(VerifyStatus.Malformed, $"non-canonicalizable: {e.Message}");
            }

            if (Sha256Prefixed(hi) != ChainHash(r))
                return new VerifyResult(VerifyStatus.Tampered, $"hash mismatch at seq {s}");

            string? aid = AgentId(r);
            string kid = SigKid(r);
            if (aid is not null)
            {
                if (pinned.TryGetValue(aid, out string? pk))
                {
                    if (pk != kid)
                        return new VerifyResult(VerifyStatus.Tampered, $"key swap for agent \"{aid}\" at seq {s}");
                }
                else
                {
                    pinned[aid] = kid;
                }
            }

            if (haveKeyring)
            {
                JVal? pub = keyring!.Get(kid);
                if (pub is not JStr pubStr)
                    return new VerifyResult(VerifyStatus.Tampered, $"unknown kid {kid} at seq {s}");
                byte[] msg = Concat(ReceiptDomain, Sha256Digest(hi));
                bool sigOk;
                try
                {
                    byte[] sigBytes = Crypto.StrictB64Decode(SigValue(r));
                    byte[] rawKey = Crypto.SpkiToRaw(pubStr.Value);
                    sigOk = Crypto.Ed25519Verify(rawKey, msg, sigBytes);
                }
                catch (Exception)
                {
                    return new VerifyResult(VerifyStatus.Tampered, $"bad signature/key encoding at seq {s}");
                }
                if (!sigOk)
                    return new VerifyResult(VerifyStatus.Tampered, $"invalid signature at seq {s}");

                if (identity is not null && !Authorized(identity, aid, kid))
                    return new VerifyResult(VerifyStatus.Untrusted, $"agent \"{aid}\" not authorized for kid \"{kid}\" at seq {s}");
            }

            // Linkage.
            JVal? link = PrevHash(r);
            if (s == 0)
            {
                if (link is not JNull)
                    return new VerifyResult(VerifyStatus.Tampered, "genesis prevHash must be null");
            }
            else
            {
                string? linkStr = link is JStr ls ? ls.Value : null;
                if (linkStr != ChainHash(prev!))
                    return new VerifyResult(VerifyStatus.Tampered, $"broken linkage at seq {s}");
            }
            prev = r;
        }

        JObj head = bySeq[receipts.Count - 1];
        var warnings = new List<string>();

        if (checkpointVal is not null)
        {
            CpVerdict cpv = VerifyCheckpoint(checkpointVal, keyring);
            if (cpv == CpVerdict.Bad)
                return new VerifyResult(VerifyStatus.Tampered, "checkpoint invalid");
            if (haveKeyring && cpv != CpVerdict.Ok)
                return new VerifyResult(VerifyStatus.Tampered, "checkpoint not authenticated against keyring");

            var cp = (JObj)checkpointVal;
            if (StrOf(cp, "chain") != chainId)
                return new VerifyResult(VerifyStatus.Tampered, "checkpoint chain mismatch");

            long cpSeq = cp.Get("highestSeq") is JInt hsq ? hsq.Value : -1;
            string? cpHead = StrOf(cp, "headHash");
            if (cpSeq != SeqOf(head) || cpHead != ChainHash(head))
                return new VerifyResult(VerifyStatus.Tampered, "chain head does not match checkpoint (tail truncated/extended)");

            if (haveKeyring && identity is not null)
            {
                JObj genesis = bySeq[0];
                string cpKid = SigKid(cp);
                if (!Authorized(identity, AgentId(genesis), cpKid))
                    return new VerifyResult(VerifyStatus.Untrusted, "checkpoint kid not authorized for chain opener (genesis) agent");

                var agents = new HashSet<string?>();
                foreach (JObj r in receipts) agents.Add(AgentId(r));
                if (agents.Count > 1)
                    warnings.Add("checkpoint completeness is opener-scoped: chain has >1 agent.id, a co-agent's tail is NOT separately certified");
            }
        }

        string detail = $"{receipts.Count} receipts, chain {chainId}";
        if (warnings.Count > 0)
            detail += " | " + string.Join(" | ", warnings);
        return new VerifyResult(haveKeyring ? VerifyStatus.Valid : VerifyStatus.Unverified, detail);
    }

    // ── field accessors (structural validation already guarantees presence/shape) ──
    private static string? ChainOf(JObj r) =>
        r.Get("scope") is JObj sc && sc.Get("chain") is JStr c ? c.Value : null;

    /// scope.tenant when present as a string, else null (ABSENT is distinct from any present value).
    private static string? TenantOf(JObj r) =>
        r.Get("scope") is JObj sc && sc.Get("tenant") is JStr t ? t.Value : null;

    private static long SeqOf(JObj r) =>
        r.Get("chain") is JObj ch && ch.Get("seq") is JInt s ? s.Value : -1;

    private static string? ChainHash(JObj r) =>
        r.Get("chain") is JObj ch && ch.Get("hash") is JStr h ? h.Value : null;

    private static JVal? PrevHash(JObj r) =>
        r.Get("chain") is JObj ch ? ch.Get("prevHash") : null;

    private static string? AgentId(JObj r) =>
        r.Get("agent") is JObj a && a.Get("id") is JStr id ? id.Value : null;

    private static string SigKid(JObj r) =>
        r.Get("sig") is JObj s && s.Get("kid") is JStr k ? k.Value : "";

    private static string SigValue(JObj r) =>
        r.Get("sig") is JObj s && s.Get("value") is JStr v ? v.Value : "";

    private static string? StrOf(JObj o, string key) => o.Get(key) is JStr s ? s.Value : null;

    private static byte[] Concat(byte[] a, byte[] b)
    {
        var r = new byte[a.Length + b.Length];
        Buffer.BlockCopy(a, 0, r, 0, a.Length);
        Buffer.BlockCopy(b, 0, r, a.Length, b.Length);
        return r;
    }

    private static string ToHex(byte[] data)
    {
        var sb = new StringBuilder(data.Length * 2);
        foreach (byte b in data) sb.Append(b.ToString("x2"));
        return sb.ToString();
    }
}

File: impl-rust/src/verify.rs
//! NOA receipt-chain verification — verdict-equivalent to impl-py `verify_chain` / `src/verify.ts`.
//! Order of checks is preserved exactly (structural → partition/dup-seq → seq-walk[hash, key-continuity,
//! signature, identity, linkage] → checkpoint tail-truncation + genesis-bound checkpoint identity), because
//! the order determines which verdict class wins when an input violates several rules at once.

use crate::jcs::canonicalize;
use crate::json::{Json, SAFE_INT_MAX};
use crate::keys::verify_sig;
use crate::schema::{is_hash, is_rfc3339_instant, validate_receipt_shape};
use sha2::{Digest, Sha256};
use std::collections::HashMap;

const RECEIPT_DOMAIN: &[u8] = b"NOA-Receipt-v0.1-sig:";
const CHECKPOINT_DOMAIN: &[u8] = b"NOA-Checkpoint-v0.1-sig:";

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Status {
    Valid,
    Unverified,
    Tampered,
    Malformed,
    Untrusted,
}

impl Status {
    pub fn exit_code(self) -> i32 {
        match self {
            Status::Valid => 0,
            Status::Unverified => 1,
            Status::Tampered => 2,
            Status::Malformed => 3,
            Status::Untrusted => 5,
        }
    }
    pub fn label(self) -> &'static str {
        match self {
            Status::Valid => "VALID",
            Status::Unverified => "UNVERIFIED",
            Status::Tampered => "TAMPERED",
            Status::Malformed => "MALFORMED",
            Status::Untrusted => "UNTRUSTED",
        }
    }
}

fn hex_lower(b: &[u8]) -> String {
    let mut s = String::with_capacity(b.len() * 2);
    for x in b {
        s.push_str(&format!("{x:02x}"));
    }
    s
}

fn sha256(data: &[u8]) -> [u8; 32] {
    let mut h = Sha256::new();
    h.update(data);
    let out = h.finalize();
    let mut r = [0u8; 32];
    r.copy_from_slice(&out);
    r
}

fn sha256_hex_prefixed(s: &str) -> String {
    format!("sha256:{}", hex_lower(&sha256(s.as_bytes())))
}

/// JCS hash input for a receipt: canonicalize with `chain.hash` and `sig.value` removed.
fn receipt_hash_input(r: &Json) -> String {
    let mut clone = r.clone();
    if let Json::Object(fields) = &mut clone {
        for (k, v) in fields.iter_mut() {
            if k == "chain" {
                if let Json::Object(cf) = v {
                    cf.retain(|(ck, _)| ck != "hash");
                }
            } else if k == "sig" {
                if let Json::Object(sf) = v {
                    sf.retain(|(sk, _)| sk != "value");
                }
            }
        }
    }
    canonicalize(&clone)
}

/// JCS hash input for a checkpoint: canonicalize with `sig.value` removed.
fn checkpoint_hash_input(cp: &Json) -> String {
    let mut clone = cp.clone();
    if let Json::Object(fields) = &mut clone {
        for (k, v) in fields.iter_mut() {
            if k == "sig" {
                if let Json::Object(sf) = v {
                    sf.retain(|(sk, _)| sk != "value");
                }
            }
        }
    }
    canonicalize(&clone)
}

fn authorized(identity: &Json, agent_id: &str, kid: &str) -> bool {
    match identity.get(agent_id) {
        Some(Json::Array(kids)) => kids.iter().any(|k| k.as_str() == Some(kid)),
        _ => false,
    }
}

/// Strict checkpoint validation + authentication. Returns "ok" / "unverified" / "bad".
/// (Any non-valid keyring entry collapses to "unverified" — with a keyring that becomes TAMPERED and
/// without one it can never reach "ok", so the final verdict is identical either way.)
fn verify_checkpoint(cp: &Json, keyring: Option<&Json>) -> &'static str {
    let obj = match cp {
        Json::Object(o) => o,
        _ => return "bad",
    };
    for (k, _) in obj {
        if !["spec", "chain", "highestSeq", "headHash", "ts", "sig"].contains(&k.as_str()) {
            return "bad";
        }
    }
    if cp.get("spec").and_then(|v| v.as_str()) != Some("noa.checkpoint/0.1") {
        return "bad";
    }
    match cp.get("chain") {
        Some(Json::Str(s)) if !s.is_empty() => {}
        _ => return "bad",
    }
    match cp.get("highestSeq").and_then(|v| v.as_int()) {
        Some(i) if (0..=SAFE_INT_MAX).contains(&i) => {}
        _ => return "bad",
    }
    match cp.get("headHash") {
        Some(Json::Str(s)) if is_hash(s) => {}
        _ => return "bad",
    }
    match cp.get("ts") {
        Some(Json::Str(s)) if is_rfc3339_instant(s) => {}
        _ => return "bad",
    }
    let sig = match cp.get("sig") {
        Some(sig @ Json::Object(so)) => {
            for (k, _) in so {
                if !["alg", "kid", "value"].contains(&k.as_str()) {
                    return "bad";
                }
            }
            sig
        }
        _ => return "bad",
    };
    if sig.get("alg").and_then(|v| v.as_str()) != Some("ed25519") {
        return "bad";
    }
    let kid = match sig.get("kid") {
        Some(Json::Str(s)) if !s.is_empty() => s.as_str(),
        _ => return "bad",
    };
    let value = match sig.get("value") {
        Some(Json::Str(s)) if !s.is_empty() => s.as_str(),
        _ => return "bad",
    };
    let pub_b64 = match keyring.and_then(|k| k.get(kid)) {
        Some(Json::Str(p)) if !p.is_empty() => p.as_str(),
        _ => return "unverified",
    };
    let hi = checkpoint_hash_input(cp);
    let mut msg = CHECKPOINT_DOMAIN.to_vec();
    msg.extend_from_slice(&sha256(hi.as_bytes()));
    match verify_sig(pub_b64, &msg, value) {
        Ok(true) => "ok",
        _ => "bad",
    }
}

/// Verify a receipt chain. Pure, offline, deterministic. Never panics — malformed shape → MALFORMED.
pub fn verify_chain(
    receipts: &Json,
    keyring: Option<&Json>,
    identity: Option<&Json>,
    checkpoint: Option<&Json>,
) -> (Status, String) {
    let arr = match receipts {
        Json::Array(a) if !a.is_empty() => a,
        _ => return (Status::Malformed, "input is not a non-empty array".into()),
    };
    if let Some(k) = keyring {
        if !k.is_object() {
            return (
                Status::Malformed,
                "keyring must be an object (kid -> base64 SPKI)".into(),
            );
        }
    }
    if let Some(id) = identity {
        match id {
            Json::Object(entries) => {
                for (aid, kids) in entries {
                    let ok = matches!(kids, Json::Array(list) if list.iter().all(|k| matches!(k, Json::Str(_))));
                    if !ok {
                        return (
                            Status::Malformed,
                            format!("identityManifest[\"{aid}\"] must be an array of kid strings"),
                        );
                    }
                }
            }
            _ => {
                return (
                    Status::Malformed,
                    "identityManifest must be an object (agent.id -> kid[])".into(),
                )
            }
        }
    }

    // Step 1: structural validation of every element, BEFORE any hashing.
    for (idx, r) in arr.iter().enumerate() {
        if let Err(e) = validate_receipt_shape(r) {
            return (Status::Malformed, format!("receipt[{idx}]: {e}"));
        }
    }

    let have_keyring = keyring.is_some();
    let chain_id = arr[0]
        .get("scope")
        .and_then(|s| s.get("chain"))
        .and_then(|c| c.as_str())
        .unwrap_or("")
        .to_string();

    // Step 2/3: single partition + unique seq.
    let mut by_seq: HashMap<i64, usize> = HashMap::new();
    for (idx, r) in arr.iter().enumerate() {
        let rc = r
            .get("scope")
            .and_then(|s| s.get("chain"))
            .and_then(|c| c.as_str())
            .unwrap_or("");
        if rc != chain_id {
            return (Status::Tampered, "multiple chain partitions".into());
        }
        let seq = r
            .get("chain")
            .and_then(|c| c.get("seq"))
            .and_then(|v| v.as_int())
            .unwrap_or(-1);
        if by_seq.contains_key(&seq) {
            return (Status::Tampered, format!("duplicate seq {seq}"));
        }
        by_seq.insert(seq, idx);
    }

    // Chain-wide scope.tenant consistency (fail-closed; matches the TS reference DEFAULT, impl-py
    // and impl-go). scope.tenant sits beside scope.chain but was enforced nowhere, so a chain mixing
    // tenants verified clean everywhere. Tenant isolation is a security boundary; the partition
    // split's verdict class is the right one. Walked in SEQ order, and a missing field is distinct
    // from a present one (absence is consistency only if it is consistent).
    {
        let tenant_of = |idx: usize| -> Option<&str> {
            arr[idx].get("scope").and_then(|s| s.get("tenant")).and_then(|t| t.as_str())
        };
        // Only a present->DIFFERENT-present drift is a cross-tenant splice. absent<->present
        // is a producer-version change on an OPTIONAL field, not tampering (see src/verify.ts).
        //
        // The comparison is NOT adjacent: comparing neighbours let an omission RESET the boundary,
        // so `acme -> globex` was TAMPERED while `acme -> absent -> globex` — the same splice with
        // one optional field left out in between — was VALID in all five implementations. The last
        // PRESENT tenant is carried across absences instead. Absence stays non-fatal; it just no
        // longer erases what the chain already committed to.
        let total = arr.len() as i64;
        let mut last_present: Option<String> = None;
        for s in 0..total {
            let cur = match by_seq.get(&s) {
                Some(c) => *c,
                None => break, // a gap is reported by the seq-walk below; do not pre-empt it
            };
            let cur_tenant = match tenant_of(cur) {
                Some(t) => t.to_string(),
                None => continue, // absence never resets the boundary, and is never fatal
            };
            if let Some(prev_present) = &last_present {
                if &cur_tenant != prev_present {
                    return (Status::Tampered, "cross-tenant splice".into());
                }
            }
            last_present = Some(cur_tenant);
        }
    }

    // Step 4: seq-walk.
    let n = arr.len();
    let mut pinned: HashMap<String, String> = HashMap::new();
    let mut prev_hash: Option<String> = None;
    for s in 0..n as i64 {
        let idx = match by_seq.get(&s) {
            Some(i) => *i,
            None => return (Status::Tampered, format!("seq gap: missing {s}")),
        };
        let r = &arr[idx];

        // 4a. hash integrity
        let hi = receipt_hash_input(r);
        let stored_hash = r
            .get("chain")
            .and_then(|c| c.get("hash"))
            .and_then(|v| v.as_str())
            .unwrap_or("");
        if sha256_hex_prefixed(&hi) != stored_hash {
            return (Status::Tampered, format!("hash mismatch at seq {s}"));
        }

        let aid = r
            .get("agent")
            .and_then(|a| a.get("id"))
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        let kid = r
            .get("sig")
            .and_then(|sg| sg.get("kid"))
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();

        // 4b. key continuity per agent.id
        if let Some(pk) = pinned.get(&aid) {
            if pk != &kid {
                return (
                    Status::Tampered,
                    format!("key swap for agent \"{aid}\" at seq {s}"),
                );
            }
        } else {
            pinned.insert(aid.clone(), kid.clone());
        }

        // 4c. signature (any missing / non-string / bad key or sig → TAMPERED, mirroring impl-py)
        if have_keyring {
            let kr = keyring.unwrap();
            let pub_b64 = match kr.get(&kid) {
                Some(Json::Str(p)) if !p.is_empty() => p.as_str(),
                _ => return (Status::Tampered, format!("unknown/invalid kid {kid} at seq {s}")),
            };
            let sig_value = r
                .get("sig")
                .and_then(|sg| sg.get("value"))
                .and_then(|v| v.as_str())
                .unwrap_or("");
            let mut msg = RECEIPT_DOMAIN.to_vec();
            msg.extend_from_slice(&sha256(hi.as_bytes()));
            match verify_sig(pub_b64, &msg, sig_value) {
                Ok(true) => {}
                _ => return (Status::Tampered, format!("invalid signature at seq {s}")),
            }

            // 4c-bis. identity binding (only meaningful once authenticated)
            if let Some(id) = identity {
                if !authorized(id, &aid, &kid) {
                    return (
                        Status::Untrusted,
                        format!("agent \"{aid}\" not authorized for kid \"{kid}\" at seq {s}"),
                    );
                }
            }
        }

        // 4d. linkage
        let link = r.get("chain").and_then(|c| c.get("prevHash"));
        if s == 0 {
            if !matches!(link, Some(Json::Null)) {
                return (Status::Tampered, "genesis prevHash must be null".into());
            }
        } else if link.and_then(|v| v.as_str()) != prev_hash.as_deref() {
            return (Status::Tampered, format!("broken linkage at seq {s}"));
        }
        prev_hash = Some(stored_hash.to_string());
    }

    // Step 5: tail-truncation via checkpoint.
    let head = &arr[*by_seq.get(&((n - 1) as i64)).unwrap()];
    if let Some(cp) = checkpoint {
        let cpv = verify_checkpoint(cp, keyring);
        if cpv == "bad" {
            return (Status::Tampered, "checkpoint invalid".into());
        }
        if have_keyring && cpv != "ok" {
            return (
                Status::Tampered,
                "checkpoint not authenticated against keyring".into(),
            );
        }
        if cp.get("chain").and_then(|v| v.as_str()).unwrap_or("") != chain_id {
            return (Status::Tampered, "checkpoint chain mismatch".into());
        }
        let cp_high = cp.get("highestSeq").and_then(|v| v.as_int());
        let head_seq = head
            .get("chain")
            .and_then(|c| c.get("seq"))
            .and_then(|v| v.as_int());
        let cp_head = cp.get("headHash").and_then(|v| v.as_str());
        let head_hash = head
            .get("chain")
            .and_then(|c| c.get("hash"))
            .and_then(|v| v.as_str());
        if cp_high != head_seq || cp_head != head_hash {
            return (
                Status::Tampered,
                "chain head does not match checkpoint (tail truncated/extended)".into(),
            );
        }
        // 5b. checkpoint identity binding to the chain OPENER (genesis, seq 0).
        if let (true, Some(id)) = (have_keyring, identity) {
            let genesis = &arr[*by_seq.get(&0).unwrap()];
            let gaid = genesis
                .get("agent")
                .and_then(|a| a.get("id"))
                .and_then(|v| v.as_str())
                .unwrap_or("");
            let cp_kid = cp
                .get("sig")
                .and_then(|sg| sg.get("kid"))
                .and_then(|v| v.as_str())
                .unwrap_or("");
            if !authorized(id, gaid, cp_kid) {
                return (
                    Status::Untrusted,
                    "checkpoint kid not authorized for chain opener (genesis) agent".into(),
                );
            }
        }
    }

    let status = if have_keyring {
        Status::Valid
    } else {
        Status::Unverified
    };
    (status, format!("{n} receipts, chain {chain_id}"))
}
